src/passes/Monomorphize.cpp

```diff
@@ -0,0 +1,197 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+//
+// When we see a call foo(arg1, arg2) and at least one of the arguments has a
+// more refined type than is declared in the function being called, create a
+// copy of the function with the refined type. That copy can then potentially be
+// optimized in useful ways later.
+//
+// Inlining also monomorphizes in effect. What this pass does is handle the
+// cases where inlining cannot be done.
+//
+// To see when monomorphizing makes sense, this optimizes the target function
+// both with and without the more refined types. If the refined types help then
+// the version with might remove a cast, for example. Note that while doing so
+// we keep the optimization results of the version without - there is no reason
+// to forget them since we've gone to the trouble anyhow. So this pass may have
+// the side effect of performing minor optimizations on functions. There is also
+// a variant of the pass that always monomorphizes, even when it does not seem
+// helpful, which is useful for testing, and possibly in cases where we need
+// more than just local optimizations to see the benefit - for example, perhaps
+// GUFA ends up more powerful later on.
+//
+// TODO: When we optimize we could run multiple cycles: A calls B calls C might
+//       end up with the refined+optimized B now having refined types in its
+//       call to C, which it did not have before. This is in fact the expected
+//       pattern of incremental monomorphization. Doing it in the pass could be
+//       more efficient as later cycles can focus only on what was just
+//       optimized and changed.
+// TODO: Not just direct calls? But updating vtables is complex.
+//
+
+#include "ir/cost.h"
+#include "ir/find_all.h"
+#include "ir/module-utils.h"
+#include "ir/names.h"
+#include "ir/type-updating.h"
+#include "ir/utils.h"
+#include "pass.h"
+#include "wasm-type.h"
+#include "wasm.h"
+
+namespace wasm {
+
+namespace {
+
+struct Monomorphize : public Pass {
+  // If set, we run some opts to see if monomorphization helps, and skip it if
+  // not.
+  bool onlyWhenHelpful;
+
+  Monomorphize(bool onlyWhenHelpful) : onlyWhenHelpful(onlyWhenHelpful) {}
+
+  void run(Module* module) override {
+    if (!module->features.hasGC()) {
+      return;
+    }
+
+    // TODO: parallelize
+
+    // Note the list of all functions. We'll be adding more, and do not want to
+    // operate on those.
+    std::vector<Name> funcNames;
+    ModuleUtils::iterDefinedFunctions(
+      *module, [&](Function* func) { funcNames.push_back(func->name); });
+
+    // Find the calls in each function and optimize where we can, changing them
+    // to call more refined targets.
+    for (auto name : funcNames) {
```

Comment:
We should probably go through and update the call sites (if not choose to generate further monomorphizations) in all of the new monomorphized functions as well.

Reply:
Yeah, either in a subsequent loop over just the new functions or by continuously looping over all functions in the module until we stop making changes.

This is related to that TODO, but would also help less complicated cases where we just happen to optimize a caller before its callee so the monomorphized version of the caller doesn't call the optimized callee.

Reply:
Makes sense. I'll add to the TODO.

Reply:
Did we end up with a TODO for this? I don't see one now, but maybe I'm missing it.

Reply:
I think the TODO on "multiple cycles" covers it, unless there is something I haven't understood from what you've said here?

https://github.com/WebAssembly/binaryen/pull/5238/files#diff-e3faa07d688c017f0e69bcd2c24731585177e0c09d40a2034c1ca0111157f0b9R37

Reply:
Landing this, but let me know if there's a TODO I should add later.

Reply:
Ah, solving the cycle problem would solve this problem, too, but solving this problem doesn't require solving the cycle problem. But I guess between the TODOs to solve the cycle problem and to visit functions in a more intelligent order, we probably don't need another separate TODO for this.